Network/DNS/IO.hs

```diff
@@ -26,51 +27,78 @@ module Network.DNS.IO (
 #define GHC708
 #endif
 
-import qualified Control.Monad.State as ST
+import qualified Control.Exception as E
 import qualified Data.ByteString.Builder as BB
 import qualified Data.ByteString.Char8 as BS
 import qualified Data.ByteString.Lazy.Char8 as LBS
 import Data.Char (ord)
-import Data.Conduit (($$+), ($$+-), ConduitM, (.|), runConduit)
-import Data.Conduit.Attoparsec (sinkParser)
-import qualified Data.Conduit.Binary as CB
-import Data.Conduit.Network (sourceSocket)
 import Data.IP (IPv4, IPv6)
 import Network.Socket (Socket)
+import System.IO.Error
+
 
 #if defined(WIN) && defined(GHC708)
-import Network.Socket (send)
+import Network.Socket (send, recv)
 import qualified Data.ByteString.Char8 as BS
 #else
-import Network.Socket.ByteString (sendAll)
+import Network.Socket.ByteString (sendAll, recv)
 #endif
 
-import Network.DNS.Decode.Internal (getResponse)
+import Network.DNS.Decode (decode)
 import Network.DNS.Encode (encode)
 import Network.DNS.Imports
-import Network.DNS.StateBinary (PState, initialState)
 import Network.DNS.Types
 
 ----------------------------------------------------------------
 
-sink :: ConduitM ByteString o IO (DNSMessage, PState)
-sink = sinkParser $ ST.runStateT getResponse initialState
-
 -- | Receiving DNS data from 'Socket' and parse it.
 
 receive :: Socket -> IO DNSMessage
-receive sock = fst <$> runConduit (sourceSocket sock .| sink)
+receive sock = do
+    bs <- recv sock 4096 `E.catch` \e -> E.throwIO $ NetworkFailure e
+    case decode bs of
```

Comment:
The hard-coded `4096` here at least deserves a comment.  We're expecting DNS replies to not exceed this length presumably because that's our advertised EDNS UDP buffer size?  It would probably do no harm to use a larger packet size here, the most we get is a single UDP datagram of 65535 bytes, perhaps we should attempt to read the whole thing whatever the size?

Reply:
@vdukhovni in principle, ipv6 jumbograms can be bigger than 65535 bytes, but I don't know the technicalities of whether this would happen with DNS.

Reply:
The ironic thing is that it is precisely with IPv6 that DNS is observed to have the most difficulties with dropped IP fragments. There are reports of lost data with sizes as small as >1280 bytes.  I don't think we need to worry about larger responses and jumbo IPv6, just 65535 (largest possible IPv4 UDP size) is enough.  Of course if can be sure that the EDNS(0) advertisement matches the limit here, we might also be OK.

Reply:
I would like to set:


```
maxUdpSize = 16384
minUdpSize = 512
```

then change `fromEDNS0` ensuring UDP size is one between them and use `maxUdpSize` in `receive`.